### PR TITLE
Add RoPE with FlexBert blocks

### PR DESCRIPTION
**Pull request by staghado**
This implements RoPE for BERT. See #3

It uses the rotary embedding kernel from [flash_attn](https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/layers/rotary.py). In the unpadded version, we first have to pad -> rope -> unpad, this is cheap especially for the model sizes we are considering.

The unpadded rope version still achieves better throughput than the padded one at large sequence lengths even with the pad/unpad logic which means that the rotary operation is negligible compared to the attention.
compare-seqlen-rope

### CODE REVIEW

**Review by ohallstrom — CHANGES_REQUESTED** [review]
Looks good in general to me, but please look at the comments I have made and make the corresponding fixes.

Would appreciate if we could have another set of eyes on this PR as well, might missed something as its late..:)

**Comment by ohallstrom** [comment]
Also, a question to the other reviewers: Would we want to support of RoPE based on position IDs?

**Comment by ohallstrom** [comment]
Also, it seems like @staghado has incorporated some code from #35 , we should make sure these changes align fully with #35 once it has been merged

File: src/bert_layers/attention.py
Comment (iacolippo) [reply src/bert_layers/attention.py:443]: It depends if we want to enable partial rotary like in GPT-J. Not sure about it

File: src/bert_layers/attention.py
Comment (staghado) [reply src/bert_layers/attention.py:531]: yeah it's left from debugging

File: src/bert_layers/embeddings.py
Comment (staghado) [reply src/bert_layers/embeddings.py:152]: It's not used but is added in order to have an identical interface between **SansPositionEmbeddings**  and **AbsoluteEmbeddings**. Avoids having custom logic in the model forward.

File: yamls/main/flex-bert-base.yaml
Comment (staghado) [reply yamls/main/flex-bert-base.yaml:1]: I think it should as there is an inconsistent use of lower and uppercase for true/false. For train_small i'll remove it but it shouldn't be an issue as we will have our custom data splits.

File: src/bert_layers/attention.py
Comment (ohallstrom) [reply src/bert_layers/attention.py:443]: If we want to enable partial rotary a solution is then simply just to set rotary_embed_size to None as default in https://github.com/AnswerDotAI/bert24/pull/36#discussion_r1615374459 . If None, then the embed dim should be set to attn_head_size. Else we can override it when we want to do partial rotary. That way we don't have to change the config twice if not doing partial rotary.

File: yamls/main/flex-bert-base.yaml
Comment (ohallstrom) [reply yamls/main/flex-bert-base.yaml:1]: In that case one could argue that it should be in a different PR as it has nothing to do with RoPE. But for me it is ok

File: src/bert_layers/attention.py
Comment (ohallstrom) [inline src/bert_layers/attention.py:636]: missing a transpose(1, 2) here? attn has shape (bs, n_heads, seq_len, head_dim)

File: src/bert_layers/attention.py
Comment (ohallstrom) [inline src/bert_layers/attention.py:543]: double transpose(1, 2), one should be removed 😅

File: src/bert_layers/attention.py
Comment (ohallstrom) [inline src/bert_layers/attention.py:415]: Itranspose(1,2) missingno?

File: src/bert_layers/attention.py
Comment (ohallstrom) [inline src/bert_layers/attention.py:628]: This row (and 611) could be moved out of the if and else statement as both if and else start with the same row

**Review by ohallstrom — COMMENTED** [review]
LGTM overall, just address the two comments and I believe we can merge

**Pull request merged**
